[Security Solution][DQD] Persist new fields in results storage

kapral18 opened the pull request:
Addresses #184751

## Summary

This PR addresses couple of issues:

### Main:
Persist revamped `resultsFieldMap` schema fields, namely `incompatibleFieldMappingItems`, `incompatibleFieldValueItems` and `sameFamilyFieldItems` in the `StorageResult` after index check, so that after release user can start accumulating data in these fields, while we prepare main UI changes.

### Additional:
Improve and narrow down existing in-house `EcsFlat` override type that originally comes from `@elastic/ecs` npm package, because currently it is too generic and too loose, resulting in an unnecessary conditional checks and leads to perception of impossible states most of which are refactored, cleaned and fixed in this PR.

### Screenshots
![image](https://github.com/elastic/kibana/assets/1625373/1cd13459-cf15-4026-84e8-3dea05eedf4d)
![image](https://github.com/elastic/kibana/assets/1625373/92593502-598a-439c-8c8e-fe3174ba963e)
![image](https://github.com/elastic/kibana/assets/1625373/67472930-5aee-4689-b748-44235bf4d9c0)

### How to test

1. Prepare index with invalid mapping and value fields + 1 same family field
```graphql
DELETE test-field-items

PUT test-field-items
{
  "mappings": {
    "properties": {
      "event.category": { "type": "keyword"},
      "agent.type": {"type": "constant_keyword" },
      "source.ip": {"type": "text"}
    }
  }
}

PUT test-field-items/_doc/1
{
  "@timestamp": "2016-05-23T08:05:34.853Z",
  "event.category": "behavior"
}

PUT test-field-items/_doc/2
{
  "@timestamp": "2016-05-23T08:05:34.853Z",
  "event.category": "shmehavior"
}
```  
2. Open DQD dashboard in kibana
3. Create `test-*` data-view with `test-*` index pattern
4. Select it in the sourcerer
5. Click expand button near test-field-items index 
6. Verify that you have 1 mapping + 1 value incompatible field + 1 same family field
7. Open kibana devtools 
8. Run
```graphql
GET .kibana-data-quality-dashboard-results-default/_search
{
  "size": 0,
  "query": { 
    "term": {
      "indexName": {
        "value": "test-field-items"
      }
    } 
  },
  "aggs": {
    "latest": {
      "terms": { "field": "indexName", "size": 10000 },
      "aggs": { 
        "latest_doc": { 
          "top_hits": { 
            "size": 1, 
            "sort": [{ "@timestamp": { "order": "desc" } }] 
          } 
        } 
      }
    }
  }
}
```
9. Verify that latest result contains `incompatibleFieldItems` and `sameFamilyFieldItems` of expected shape:
```json5
//...
                     "incompatibleFieldValueItems": [
                      {
                        "fieldName": "event.category",
                        "expectedValues": [
                          "api",
                          "authentication",
                          "configuration",
                          "database",
                          "driver",
                          "email",
                          "file",
                          "host",
                          "iam",
                          "intrusion_detection",
                          "library",
                          "malware",
                          "network",
                          "package",
                          "process",
                          "registry",
                          "session",
                          "threat",
                          "vulnerability",
                          "web"
                        ],
                        "actualValues": [
                          { "name": "behavior",  count: 2 },
                          { "name": "shmehavior", count: 1}
                        ],
                        "description": """This is one of four ECS Categorization Fields, and indicates the second level in the ECS category hierarchy.
`event.category` represents the "big buckets" of ECS categories. For example, filtering on `event.category:process` yields all events relating to process activity. This field is closely related to `event.type`, which is used as a subcategory.
This field is an array. This will allow proper categorization of some events that fall in multiple categories."""
                      }
                     ],
                     "incompatibleFieldMappingItems": [
                      {
                        "fieldName": "source.ip",
                        "expectedValue": "ip",
                        "actualValue": "text",
                        "description": "IP address of the source (IPv4 or IPv6)."
                      }
                    ]
//...
"sameFamilyFieldItems": [
                      {
                        "fieldName": "agent.type",
                        "expectedValue": "keyword",
                        "actualValue": "constant_keyword",
                        "description": """Type of the agent.
The agent type always stays the same and should be given by the agent used. In case of Filebeat the agent would always be Filebeat also if two Filebeat instances are run on the same machine."""
                      }
                    ]
```

kapral18 committed:
refactor: update types and refactor tests for ECS data quality dashboard

- Refactor `EnrichedFieldMetadata` into `EcsBasedFieldMetadata` and `CustomFieldMetadata`
- Update tests to use new types
- Remove unnecessary null checks for `ecsMetadata`
- Simplify rendering logic in various components
- Add `constants.ts` for ECS metadata constants

kapral18 committed:
feat(ecs_data_quality_dashboard): enhance StorageResult with detailed field items

- Added `incompatibleFieldItems` and `sameFamilyFieldItems` to `StorageResult` for detailed field information.
- Updated tests to validate the new functionality.
- Adjusted type definitions and mock data to support the changes.

Addresses #184751

elasticmachine commented:
Pinging @elastic/security-threat-hunting-explore (Team:Threat Hunting:Explore)

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/get_common_table_columns/index.tsx:
```diff
@@ -44,19 +44,25 @@ export const getCommonTableColumns = (): Array<
+        return <CodeSuccess data-test-subj="codeSuccess">{x.indexFieldType}</CodeSuccess>;
+      }
+
+      // mapping mismatch due to same family
+      if (getIsInSameFamily({ ecsExpectedType: x.type, type: x.indexFieldType })) {
+        return (
           <div>
             <CodeSuccess data-test-subj="codeSuccess">{x.indexFieldType}</CodeSuccess>
             <SameFamily />
           </div>
-        ) : (
-          <CodeDanger data-test-subj="codeDanger">{x.indexFieldType}</CodeDanger>
-        )
-      ) : (
-        <CodeSuccess data-test-subj="codeSuccess">{x.indexFieldType}</CodeSuccess>
-      ),
+        );
+      }
+
+      // mapping mismatch
+      return <CodeDanger data-test-subj="codeDanger">{x.indexFieldType}</CodeDanger>;
+    },
```
For reviewers:

`hasEcsMetadata` is now the discriminant that differentiates `EcsBasedFieldMetadata` from `CustomFieldMetadata`. Which means we no longer need to check `x.type` because it's guaranteed per the refactored type.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/get_incompatible_mappings_table_columns/index.tsx:
```diff
@@ -28,11 +28,7 @@ export const getIncompatibleMappingsTableColumns = (): Array<
     name: i18n.ECS_MAPPING_TYPE_EXPECTED,
-    render: (type: string) => (
-      <CodeSuccess data-test-subj="codeSuccess">
-        {type != null ? type : EMPTY_PLACEHOLDER}
```
we remove this check because it's an impossible state after updated types. `getIncompatibleMappingsTableColumns` only accepts `incompatible` fields which  by definition always have `type` field.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/get_incompatible_mappings_table_columns/index.test.tsx:
```diff
@@ -65,19 +65,6 @@ describe('getIncompatibleMappingsTableColumns', () => {
 
       expect(screen.getByTestId('codeSuccess')).toHaveTextContent(expected);
     });
-
-    test('it renders an empty placeholder when type is undefined', () => {
-      const columns = getIncompatibleMappingsTableColumns();
-      const typeColumnRender = columns[1].render;
-
-      render(
-        <TestProviders>
-          {typeColumnRender != null && typeColumnRender(undefined, eventCategory)}
-        </TestProviders>
-      );
-
-      expect(screen.getByTestId('codeSuccess')).toHaveTextContent(EMPTY_PLACEHOLDER);
-    });
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631423077

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/helpers.test.tsx:
```diff
@@ -117,31 +116,6 @@ describe('helpers', () => {
           expect(screen.queryByTestId('typePlaceholder')).not.toBeInTheDocument();
         });
       });
-
-      describe('when `type` is undefined', () => {
-        beforeEach(() => {
-          const withUndefinedType = {
-            ...eventCategory,
-            type: undefined, // <--
-          };
-          const columns = getEcsCompliantTableColumns();
-          const typeRender = columns[1].render;
-
-          render(
-            <TestProviders>
-              <>{typeRender != null && typeRender(withUndefinedType.type, withUndefinedType)}</>
-            </TestProviders>
-          );
-        });
-
-        test('it does NOT render the `type`', () => {
-          expect(screen.queryByTestId('type')).not.toBeInTheDocument();
-        });
-
-        test('it renders the placeholder', () => {
-          expect(screen.getByTestId('typePlaceholder')).toHaveTextContent(EMPTY_PLACEHOLDER);
-        });
-      });
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631423077

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/helpers.test.tsx:
```diff
@@ -230,35 +204,6 @@ describe('helpers', () => {
           expect(screen.queryByTestId('emptyPlaceholder')).not.toBeInTheDocument();
         });
       });
-
-      describe('when `description` is undefined', () => {
-        const withUndefinedDescription = {
-          ...eventCategory,
-          description: undefined, // <--
-        };
-
-        beforeEach(() => {
-          const columns = getEcsCompliantTableColumns();
-          const descriptionRender = columns[3].render;
-
-          render(
-            <TestProviders>
-              <>
-                {descriptionRender != null &&
-                  descriptionRender(withUndefinedDescription.description, withUndefinedDescription)}
-              </>
-            </TestProviders>
-          );
-        });
-
-        test('it does NOT render the `description`', () => {
-          expect(screen.queryByTestId('description')).not.toBeInTheDocument();
-        });
-
-        test('it renders the placeholder', () => {
-          expect(screen.getByTestId('emptyPlaceholder')).toBeInTheDocument();
-        });
-      });
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631423077

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/helpers.tsx:
```diff
@@ -52,12 +57,7 @@ export const getEcsCompliantTableColumns = (): Array<
     field: 'type',
     name: i18n.ECS_MAPPING_TYPE,
-    render: (type: string | undefined) =>
-      type != null ? (
```
For reviewers:

we remove this check because it checks an impossible state.`getEcsCompliantTableColumns` only accepts ecs based fields which by new type definition always have `type` field.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/helpers.tsx:
```diff
@@ -75,20 +75,15 @@ export const getEcsCompliantTableColumns = (): Array<
     field: 'description',
     name: i18n.ECS_DESCRIPTION,
-    render: (description: string | undefined) =>
-      description != null ? (
```
For reviewers:

we remove this check because it checks an impossible state.`getEcsCompliantTableColumns` only accepts ecs based fields which by new type definition always have `description` field.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/index.tsx:
```diff
@@ -53,4 +53,6 @@ const CompareFieldsTableComponent: React.FC<Props> = ({
 CompareFieldsTableComponent.displayName = 'CompareFieldsTableComponent';
 
-export const CompareFieldsTable = React.memo(CompareFieldsTableComponent);
+export const CompareFieldsTable = React.memo(
+  CompareFieldsTableComponent
+) as typeof CompareFieldsTableComponent;
```
For reviewers:

This is the cheapast fix for `React.memo` generic type forwarding. `React.memo` type layer is insignificant anyways so this doesn't affect type safety.

semd replied:
From what I understand, the `T` type parameter has been introduced to allow different subtypes for `enrichedFieldMetadata` and `getTableColumns`. But is the benefit worth all these "type gymnastics"? What's the practical difference between this approach, and using the generic `EnrichedFieldMetadata` type directly for those props?

kapral18 replied:
Good catch. 

My intention was to make it more consistent with the rest of the changes and a little less confusing for getCustom... getEcs... getIncompatible... methods type-wise, because they are not supposed to return a wide type, they are supposed to return what their name explicitly mentions, it will be a problem the moment you try to access specific props that are only present in explicit type. But right now it's not happening and complicates the review. So I will undo these changes. 

Thanks.

kapral18 replied:
done

kapral18 replied:
oh and while undoing the changes @semd I came across a perfect example of why this change was actually pretty beneficial.

https://github.com/elastic/kibana/blob/8cd4858d2fbde97d3513feb58ee07cca304898cb/x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/get_incompatible_mappings_table_columns/index.tsx#L44

Here this assumes that `x` is of type `EnrichedFieldMetadata`. Which means every new logic you add here or even a mere check for `x.type` will create a type error here, you will have to needlessly narrow it down, even tho practically this function will always only accept incompatible fields that DO have `type` on them as opposed to custom fields. So by not narrowing down the type here (via generics in the accepting component) you are creating a situation when impossible states will happen, because TS will force them to check or assert, both of which is a bad and unnecessary. they will also have to accompany this with a needless test most of which I cleared in this PR. Let me know if this makes sense.

kapral18 replied:
https://github.com/elastic/kibana/assets/1625373/75aaca3d-1fdd-4f83-aa7a-9af20e1e93f5

semd replied:
Okay, I see. It's funny because TS does not complain about `enrichedFieldMetadata` it detects those types are inclusive, it only complains about the `getTableColumns` I guess that's because it is a generic type. TS doing TS things....

If you want, there's another approach. Composing those components would allow us to forget about column definitions as a prop. Creating a separate component for `CompareInvalidFieldsTable`, `CompareECSFieldsTable`, `CompareCustomFieldsTable`, and `CompareSameFamilyFieldsTable`. They all would compose the current `CompareFieldsTable`, and each of them would have their own table columns definition inside the component, only receiving props for the _title_ and _fieldMetadata_. This way we also centralize all the table column definitions.

Afaik, React recommends "composing" over "extending", so this is the standard way I deal with this kind of things.
However, your first approach is also a good solution, this is not a "requested change" it's just a suggestion you can consider if you want.

kapral18 replied:
> Okay, I see. It's funny because TS does not complain about enrichedFieldMetadata it detects those types are inclusive, it only complains about the getTableColumns I guess that's because it is a generic type. TS doing TS things....

Yes because here 
![image](https://github.com/elastic/kibana/assets/1625373/0392070d-cbfc-4833-ad7d-e3e50080d1ca)

`EnrichedFieldMetadata` is including `EcsBasedFieldMetadata` and `CustomFieldMetadata`. So by passing narrower type into wider type you get no errors. And that's a logical and expected outcome. Because TS assumes that you just want to be able to pass in subtypes and that's exactly what we want here.

The problem happens inside of one of each of the methods like 
![image](https://github.com/elastic/kibana/assets/1625373/70a16f86-2915-4ca2-bcfb-9a3f853965d5)

Which is also very expected if you tell TS that "hey I will have `EnrichedFieldMetadata`" in here, so when accessing `x.<field>`, it breaks down as it SHOULD. This is a safety mechanism which is to be expected for you to not accidentally try to access a field that doesn't exist.

That's why you need to tell TS exactly what you expect to work with or as close as possible so it can help you better avoid needless checks and needless tests for those checks.

Long story short, TS is good here, it helps and defends against impossible states and confusion. The most important thing is to keep your root types healthy and as close to implementation as possible. If that's done TS will do its job of making sure all those impossible tests and checks never happen again in the future.

> If you want, there's another approach. 

Firstly I would suggest we start with why the particular composition approach (which I don't fully understand from your description unfortunately) will work better than existing suggestion from perspective type changes? Meaning what is the unfixable or cumbersome problem that current columned approach is introducing that would otherwise not  happen? Because if the problem is the introduction of component generic types, I'd argue it's not really a problem, but just another valid use of types with components. The `typeof Component` thing is just a limitation of react types nothing more, it doesn't make the entire approach invalid and needing to be completely refactored.

But I am open to your arguments and reasoning about the potential change, let's catch up in person during my implementation of the UI changes.

And now I will resolve this discussion and return my changes. Thanks for your input.

angorayc replied:
Zoomed with Karen, I am convinced that the casting here was to support the generic type for CompareFieldsTable, so we can narrow dow EnrichedFieldMetadata to the specific types when rendered in each tab. Making the types specific can reduce redundant checks in the future and therefor reduce the test cases needed to cover those logic. I'm fine with the change here.

kapral18 replied:
thanks @angorayc also as mentioned in our discussion if this use case becomes more popular we can always extract it out into util. And btw just noticed it's already used in kibana in other teams as well  ![image](https://github.com/elastic/kibana/assets/1625373/3a2406f7-1fad-4560-b297-d32e2ccb7152)

semd replied:
Yeah, I am also fine, I am not saying the current approach is wrong. I am just saying that this approach is something we don't usually do, this would be the first component I see doing that. That's why I would lean towards composing rather than making generic types, but I think both are okay. The behavior would be pretty much the same. But instead of doing this:
```
<CompareFieldsTable
    title="foo"
    enrichedFieldMetadata={incompatibleFieldsMetadata} // generic T = EcsFieldMetadata
    getTableColumns(getIncompatibleFieldsTableColumns} // Array<EuiTableFieldDataColumnType<EcsFieldMetadata>>
/>
```

We would do this:
```
<CompareIncompatibleFieldsTable
    title="foo"
    fieldsMetadata={incompatibleFieldsMetadata} // explicit EcsFieldMetadata type
/>
``` 

I like it because it's more explicit and simple, and there is some functionality (column definition) we could encapsulate inside each component. 
But as you want, I am okay with both.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/allowed_values/helpers.tsx:
```diff
@@ -5,40 +5,38 @@
+  ecsMetadata: EcsFlatTyped;
   indexName: string;
 }): UnallowedValueRequestItem[] =>
-  ecsMetadata != null
```
For reviewers:

we remove this check because it checks an impossible state. `ecsMetadata` represents `EcsFlat` object which is imported statically so it can never be null.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/allowed_values/helpers.tsx:
```diff
@@ -5,40 +5,38 @@
+  ecsMetadata: EcsFlatTyped;
   fieldName: string;
-}): boolean =>
-  ecsMetadata != null ? (ecsMetadata[fieldName]?.allowed_values?.length ?? 0) > 0 : false;
```
For reviewers:

we remove this check because it checks an impossible state. `ecsMetadata` represents `EcsFlat` object which is imported statically so it can never be null.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/allowed_values/helpers.test.tsx:
```diff
@@ -36,25 +31,16 @@ describe('helpers', () => {
         })
       ).toBe(false);
     });
-
-    test('it returns false when `ecsMetadata` is null', () => {
-      expect(
-        hasAllowedValues({
-          ecsMetadata: null, // <--
-          fieldName: 'event.category',
-        })
-      ).toBe(false);
-    });
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631497248

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/allowed_values/helpers.test.tsx:
```diff
@@ -79,60 +65,19 @@ describe('helpers', () => {
-    test('it skips `allowed_values` where `name` is undefined', () => {
-      // omit the `name` property from the `database` `AllowedValue`:
-      const missingDatabase =
-        ecsMetadata['event.category'].allowed_values?.map((x) =>
-          x.name === 'database' ? omit<AllowedValue>('name', x) : x
-        ) ?? [];
-
-      const field = {
-        ...ecsMetadata['event.category'],
-        allowed_values: missingDatabase,
-      };
-
-      expect(getValidValues(field)).toEqual(
-        expect.arrayContaining([
-          'authentication',
-          'configuration',
-          'driver',
-          'email',
-          'file',
-          'host',
-          'iam',
-          'intrusion_detection',
-          'malware',
-          'network',
-          'package',
-          'process',
-          'registry',
-          'session',
-          'threat',
-          'vulnerability',
-          'web',
-        ])
-      );
-      expect(getValidValues(field)).not.toEqual(
-        expect.arrayContaining([
-          // there should be no entry for 'database'
-          'database',
-        ])
-      );
-    });
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631495151

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/allowed_values/helpers.test.tsx:
```diff
@@ -203,14 +148,5 @@ describe('helpers', () => {
         },
       ]);
     });
-
-    test('it returns an empty array when `ecsMetadata` is null', () => {
-      expect(
-        getUnallowedValueRequestItems({
-          ecsMetadata: null, // <--
-          indexName: 'auditbeat-*',
-        })
-      ).toEqual([]);
-    });
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631493542

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/data_quality_summary/summary_actions/check_all/check_index.ts:
```diff
@@ -41,7 +42,7 @@ export async function checkIndex({
   abortController: AbortController;
   batchId: string;
   checkAllStartTime: number;
-  ecsMetadata: Record<string, EcsMetadata> | null;
```
For reviewers:

`ecsMetadata` represents `EcsFlat` object which is imported statically so it can never be `null`.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/data_quality_summary/summary_actions/check_all/check_index.test.ts:
```diff
@@ -164,51 +163,6 @@ describe('checkIndex', () => {
-
-      await checkIndex({
-        abortController: new AbortController(),
-        batchId: 'batch-id',
-        checkAllStartTime: Date.now(),
-        ecsMetadata: null, // <--
-        formatBytes,
-        formatNumber,
-        httpFetch,
-        indexName,
-        isLastCheck: false,
-        onCheckCompleted,
-        pattern,
-        version: EcsVersion,
-      });
-    });
-
-    test('it invokes onCheckCompleted with a null `error`', () => {
-      expect(onCheckCompleted.mock.calls[0][0].error).toBeNull();
-    });
-
-    test('it invokes onCheckCompleted with the expected `indexName`', () => {
-      expect(onCheckCompleted.mock.calls[0][0].indexName).toEqual(indexName);
-    });
-
-    test('it invokes onCheckCompleted with the default `partitionedFieldMetadata`', () => {
-      expect(onCheckCompleted.mock.calls[0][0].partitionedFieldMetadata).toEqual(
-        EMPTY_PARTITIONED_FIELD_METADATA
-      );
-    });
-
-    test('it invokes onCheckCompleted with the expected `pattern`', () => {
-      expect(onCheckCompleted.mock.calls[0][0].pattern).toEqual(pattern);
-    });
-
-    test('it invokes onCheckCompleted with the expected `version`', () => {
-      expect(onCheckCompleted.mock.calls[0][0].version).toEqual(EcsVersion);
-    });
-  });
-
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631554934

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/tabs/callouts/incompatible_callout/helpers.test.ts:
For reviewers: dead code

kapral18 replied:
this was cleaned by upstream changes, resolving...

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/tabs/callouts/incompatible_callout/helpers.ts:
For reviewers: dead code

kapral18 replied:
this was cleaned by upstream changes, resolving...

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/index_properties/markdown/helpers.test.ts:
```diff
@@ -230,17 +230,6 @@ describe('helpers', () => {
-    test('it returns the expected table rows when some have allowed values', () => {
-      const withAllowedValues = [
-        ...mockCustomFields,
-        eventCategory, // note: this is not a real-world use case, because custom fields don't have allowed values
```
For reviewes:

as explained here this test covers an impossible state

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/index_properties/helpers.ts:
```diff
@@ -40,19 +41,21 @@ export const getSortedPartitionedFieldMetadata = ({
   mappingsProperties: Record<string, MappingProperty> | null | undefined;
   unallowedValues: Record<string, UnallowedValueCount[]> | null;
 }): PartitionedFieldMetadata | null => {
-  if (loadingMappings || ecsMetadata == null || unallowedValues == null) {
```
For reviewers:

`ecsMetadata` represents `EcsFlat` object which is imported statically so it can never be `null`.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/index_properties/helpers.ts:
```diff
@@ -40,19 +41,21 @@ export const getSortedPartitionedFieldMetadata = ({
   mappingsProperties,
   unallowedValues,
 }: {
-  ecsMetadata: Record<string, EcsMetadata> | null;
```
For reviewers:

ecsMetadata represents EcsFlat object which is imported statically so it can never be null.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/index_properties/helpers.test.ts:
```diff
@@ -5,47 +5,32 @@
       ).toBeNull();
     });
 
-    test('it returns null when `ecsMetadata` is null', () => {
-      expect(
-        getSortedPartitionedFieldMetadata({
-          ecsMetadata: null, // <--
-          loadingMappings: false,
-          mappingsProperties: mockMappingsProperties,
-          unallowedValues: {},
-        })
-      ).toBeNull();
-    });
-
```
For reviewers:

This test was removed because it tests an impossible state. Details:
https://github.com/elastic/kibana/pull/185025/files#r1631679640

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/index_properties/helpers.test.ts:
```diff
@@ -54,30 +39,27 @@ describe('helpers', () => {
-            indexInvalidValues: [],
-            isEcsCompliant: false,
-            isInSameFamily: false,
-            type: 'date',
-          },
-        ],
+        incompatible: [incompatibleFieldMetadata],
```
For reviewers:

Per logic of `getPartitionedFieldMetadata` that is used within `getSortedFieldMetadata`, `all` is should always include `incompatible` and the rest

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/tabs/helpers.test.tsx:
```diff
@@ -38,11 +37,11 @@ describe('helpers', () => {
     test('it returns false when `enrichedFieldMetadata` contains an @timestamp field', () => {
-      expect(showMissingTimestampCallout([timestamp, eventCategory, someField])).toBe(false);
+      expect(showMissingTimestampCallout([timestamp, eventCategory])).toBe(false);
     });
 
     test('it returns true when `enrichedFieldMetadata` does NOT contain an @timestamp field', () => {
-      expect(showMissingTimestampCallout([eventCategory, someField])).toBe(true);
+      expect(showMissingTimestampCallout([eventCategory])).toBe(true);
```
For reviewers:

After narroweding down types turns out only `ecsBasedFieldMetadata` is and should be passed into `showMissingTimestampCallout`, hence `someField` being a custom field mock and yielding a type error is getting removed

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/helpers.test.ts:
```diff
@@ -717,37 +713,6 @@ describe('helpers', () => {
     });
   });
 
-  describe('hasValidTimestampMapping', () => {
-    test('it returns true when the `enrichedFieldMetadata` has a valid @timestamp', () => {
-      const enrichedFieldMetadata: EnrichedFieldMetadata[] = [timestamp, sourcePort];
-
-      expect(hasValidTimestampMapping(enrichedFieldMetadata)).toBe(true);
-    });
-
-    test('it returns false when the `enrichedFieldMetadata` collection does NOT include a valid @timestamp', () => {
-      const enrichedFieldMetadata: EnrichedFieldMetadata[] = [sourcePort];
-
-      expect(hasValidTimestampMapping(enrichedFieldMetadata)).toBe(false);
-    });
-
-    test('it returns false when the `enrichedFieldMetadata` has an @timestamp with an invalid mapping', () => {
-      const timestampWithInvalidMapping: EnrichedFieldMetadata = {
-        ...timestamp,
-        indexFieldType: 'text', // invalid mapping, should be "date"
-      };
-      const enrichedFieldMetadata: EnrichedFieldMetadata[] = [
-        timestampWithInvalidMapping,
-        sourcePort,
-      ];
-
-      expect(hasValidTimestampMapping(enrichedFieldMetadata)).toBe(false);
-    });
-
-    test('it returns false when `enrichedFieldMetadata` is empty', () => {
-      expect(hasValidTimestampMapping([])).toBe(false);
-    });
-  });
-
```
For reviewers: dead code

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/helpers.ts:
```diff
@@ -258,11 +260,6 @@ export const getPartitionedFieldMetadataStats = (
   };
 };
 
-export const hasValidTimestampMapping = (enrichedFieldMetadata: EnrichedFieldMetadata[]): boolean =>
-  enrichedFieldMetadata.some(
-    (x) => x.indexFieldName === '@timestamp' && x.indexFieldType === 'date'
-  );
-
```
For reviewers: dead code

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/helpers.ts:
```diff
@@ -202,23 +205,22 @@ export const getEnrichedFieldMetadata = ({
       indexFieldName: field,
       indexFieldType: type,
-      indexInvalidValues,
+      indexInvalidValues: [],
```
For reviewers: 

Just making it a little more explicit that this branch of logic only has 2 dynamic values `field` and `type` the rest is hard coded and constitutes a separate `CustomFieldMetadata` type with a discriminant in `hasEcsMetadata` field

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/types.ts:
```diff
@@ -16,43 +16,73 @@ export interface Mappings {
-  description?: string;
-  example?: string;
-  flat_name?: string;
+  beta?: string;
+  doc_values?: boolean;
+  example?: string | number | boolean;
+  expected_values?: string[];
   format?: string;
   ignore_above?: number;
-  level?: string;
-  name?: string;
-  normalize?: string[];
+  index?: boolean;
+  input_format?: string;
+  multi_fields?: MultiField[];
+  object_type?: string;
+  original_fieldset?: string;
+  output_format?: string;
+  output_precision?: number;
+  pattern?: string;
   required?: boolean;
-  short?: string;
-  type?: string;
+  scaling_factor?: number;
+}
+
+export interface AllowedValue {
+  description: string;
+  name: string;
+  expected_event_types?: string[];
+  beta?: string;
 }
 
-export type EnrichedFieldMetadata = EcsMetadata & {
-  hasEcsMetadata: boolean;
+export interface MultiField {
+  flat_name: string;
+  name: string;
+  type: string;
+}
```
For reviewers:

This is a precise cumulative type inference of a given field from `@elastic/ecs` npm package. All required fields are actually required in `ecs_flat.js` file that is imported and used in our code.

angorayc reviewed (approved):
Tested locally, all worked as expected.

kapral18 commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/index_properties/helpers.test.ts:
```diff
@@ -54,30 +39,27 @@ describe('helpers', () => {
     });
 
     describe('when `mappingsProperties` is unknown', () => {
+      const incompatibleFieldMetadata = {
```
@angorayc I think current naming is correct because when `mappingProperies` is `null` or `undefined` the idea is to return an incompatible field to show it as a result of a check. You can see in the test title further down below that incompatible is used many times on purpose. It's also true that we add this into the collection of incompatible fieldMetadata explicitly when returning partitionedMetadata.

For example If I create an empty index without any mapping and try to check it I get back an incompatible field as a result of check denoting that `@timestamp` is invalid, as in - it's required but not present and it should be.

![image](https://github.com/elastic/kibana/assets/1625373/df5e8033-3098-4a2f-a4e4-890b0570a2f7)

angorayc replied:
I see, actually in this scenario it's valid that the `all` property contains the same data as `incompatible`, and it's indeed incompatibleFieldMetadata. Got it, I'm fine with it, thank you!

andrew-goldstein commented on x-pack/plugins/ecs_data_quality_dashboard/server/schemas/result.ts:
```diff
@@ -19,8 +19,25 @@ const ResultDocumentInterface = t.interface({
+    t.type({
+      fieldName: t.string,
+      expectedValue: t.union([t.string, t.array(t.string)]),
+      actualValue: t.union([t.string, t.array(t.string)]),
```
Consider storing the counts of the incompatible values (in addition to the incompatible values) to make it possible to render (from stored results) the counts shown in the screenshot below:

![count_of_actual_value](https://github.com/elastic/kibana/assets/4459398/a0e1df6f-17a1-474c-9906-a6f820302002)

To create the screenshot above, a 2nd document containing the value was inserted (via dev tools): 

```
PUT test-field-items/_doc/2
{
  "@timestamp": "2024-06-11T08:05:34.853Z",
  "event.category": "behavior"
}
```

The count is already (correctly) included in the `markdownComments` field in stored results: 

```json
                {
                  "_index": ".ds-.kibana-data-quality-dashboard-results-default-2024.05.22-000001",
                  "_id": "PuIGCZABPzDWWqs-bmTR",
                  "_score": null,
                  "_source": {
                    "batchId": "a650d1f6-c0d7-400a-b023-23edd23e27af",
                    "indexName": "test-field-items",
                    "indexPattern": "test-*",
                    "isCheckAll": false,
                    "checkedAt": 1718138334792,
                    "docsCount": 1,
                    "totalFieldCount": 4,
                    "ecsFieldCount": 1,
                    "customFieldCount": 0,
                    "incompatibleFieldCount": 2,
                    "incompatibleFieldItems": [
                      {
                        "fieldName": "event.category",
                        "expectedValue": [
                          "api",
                          "authentication",
                          "configuration",
                          "database",
                          "driver",
                          "email",
                          "file",
                          "host",
                          "iam",
                          "intrusion_detection",
                          "library",
                          "malware",
                          "network",
                          "package",
                          "process",
                          "registry",
                          "session",
                          "threat",
                          "vulnerability",
                          "web"
                        ],
                        "actualValue": [
                          "behavior"
                        ],
                        "description": """This is one of four ECS Categorization Fields, and indicates the second level in the ECS category hierarchy.
`event.category` represents the "big buckets" of ECS categories. For example, filtering on `event.category:process` yields all events relating to process activity. This field is closely related to `event.type`, which is used as a subcategory.
This field is an array. This will allow proper categorization of some events that fall in multiple categories.""",
                        "reason": "value"
                      },
                      {
                        "fieldName": "source.ip",
                        "expectedValue": "ip",
                        "actualValue": "text",
                        "description": "IP address of the source (IPv4 or IPv6).",
                        "reason": "mapping"
                      }
                    ],
                    "sameFamilyFieldCount": 1,
                    "sameFamilyFields": [
                      "agent.type"
                    ],
                    "sameFamilyFieldItems": [
                      {
                        "fieldName": "agent.type",
                        "expectedValue": "keyword",
                        "actualValue": "constant_keyword",
                        "description": """Type of the agent.
The agent type always stays the same and should be given by the agent used. In case of Filebeat the agent would always be Filebeat also if two Filebeat instances are run on the same machine."""
                      }
                    ],
                    "unallowedMappingFields": [
                      "source.ip"
                    ],
                    "unallowedValueFields": [
                      "event.category"
                    ],
                    "sizeInBytes": 8780,
                    "ilmPhase": "unmanaged",
                    "markdownComments": [
                      """### test-field-items
""",
                      """| Result | Index | Docs | Incompatible fields | ILM Phase | Size |
|--------|-------|------|---------------------|-----------|------|
| ❌ | test-field-items | 1 (100.0%) | 2 | `unmanaged` | 8.6KB |

""",
                      """### **Incompatible fields** `2` **Same family** `1` **Custom fields** `0` **ECS compliant fields** `1` **All fields** `4`
""",
                      """#### 2 incompatible fields

Fields are incompatible with ECS when index mappings, or the values of the fields in the index, don't conform to the Elastic Common Schema (ECS), version 8.11.0.

❌ Detection engine rules referencing these fields may not match them correctly
❌ Pages may not display some events or fields due to unexpected field mappings or values
❌ Mappings or field values that don't comply with ECS are not supported
""",
                      """
#### Incompatible field mappings - test-field-items


| Field | ECS mapping type (expected) | Index mapping type (actual) | 
|-------|-----------------------------|-----------------------------|
| source.ip | `ip` | `text`  |

#### Incompatible field values - test-field-items


| Field | ECS values (expected) | Document values (actual) | 
|-------|-----------------------|--------------------------|
| event.category | `api`, `authentication`, `configuration`, `database`, `driver`, `email`, `file`, `host`, `iam`, `intrusion_detection`, `library`, `malware`, `network`, `package`, `process`, `registry`, `session`, `threat`, `vulnerability`, `web` | `behavior` (2) |

"""
                    ],
                    "ecsVersion": "8.11.0",
                    "indexId": "NoMNexi_Th-duyhPmnDmlA",
                    "error": null,
                    "@timestamp": 1718138334831,
                    "checkedBy": "u_mGBROF_q5bmFCATbLXAcCwKa0k8JvONAwSruelyKA5E_0"
                  },
                  "sort": [
                    1718138334831
                  ]
                }
```

semd replied:
Good catch @andrew-goldstein, we did not realize this counter existed. 
@kapral18 what do you think about adding `actualValueCount` inside the `incompatibleFieldItems`? I assume it's also necessary for `sameFamilyFieldItems` as well.

kapral18 replied:
@andrew-goldstein I knew the wait was worth the while, thanks for finding this.

@semd 2 things to keep in mind:

1. SameFamily fields are valid so cannot have actual value counts, they only show up if there is a type mismatch within acceptable match, so I guess we don't need to have any of that in sameFamilyFieldItems
![image](https://github.com/elastic/kibana/assets/1625373/09b4f77e-a8f3-40a1-be80-af85fb177d7a)

2. Since we can have multiple actualValues for the same field 
![image](https://github.com/elastic/kibana/assets/1625373/26ce6ccf-45e6-4644-8d47-ea9368ed3c1f)
I'd argue it's a little more convenient to include the count into `actualValue` list but now I am also thinking about tweaking the shape a little more to make it more evident that it's an actual list in case of incompatible items vs same family field to avoid confusions like this.

I will push the change and it will be easier to see the suggestion in the separate commit.

semd replied:
@kapral18 correct, "same family" is always a mapping issue, they don't have a value appearance counter. Same for "mapping incompatibilities", this counter is only needed for "value incompatibilities". This will need to be stored per-"incompatible actual value".

kapral18 replied:
@andrew-goldstein @semd @angorayc after weighing the options I decided to revamp the underlying structure one last time. I believe it's much more explicit now. Please have a look.

The gist of changes I separate incompatible mappings and values from each other and values now contain `actualValues` instead of `actualValue` and include the `count` with each invalid value item

kapral18 replied:
I also updated the PR description

andrew-goldstein reviewed (approved):
Thanks @kapral18 for the enhancements to persistence, and the improved types! 🙏 
✅ Desk tested locally
LGTM 🚀

kapral18 committed:
refactor: split incompatibleFieldItems into mapping and value items

- Refactored `incompatibleFieldItems` into `incompatibleFieldMappingItems` and `incompatibleFieldValueItems`.
- Updated related test cases in `helpers.test.ts`.
- Modified `helpers.ts` to handle the new structure.
- Updated type definitions in `types.ts`.
- Adjusted `results_field_map.ts` to reflect the new structure.
- Updated mock data in `results.mock.ts`.
- Modified schema definitions in `result.ts`.
